libkbfs/journal_server.go

```diff
@@ -176,18 +218,130 @@ func (j *JournalServer) Enable(ctx context.Context, tlfID TlfID) (err error) {
 		return err
 	}
 
-	bundle.blockJournal = blockJournal
 	mdJournal, err := makeMDJournal(
 		uid, key, j.config.Codec(), j.config.Crypto(), tlfDir, log)
 	if err != nil {
 		return err
 	}
 
-	bundle.mdJournal = mdJournal
+	bundle := makeTlfJournalBundle(blockJournal, mdJournal)
 	j.tlfBundles[tlfID] = bundle
+
+	go j.autoFlush(tlfID, afs, bundle.hasWorkCh, bundle.pauseCh,
+		bundle.resumeCh, bundle.shutdownCh)
+
+	// Signal work to pick up any existing journal entries.
+	select {
+	case bundle.hasWorkCh <- struct{}{}:
+	default:
+	}
+
 	return nil
 }
 
+func (j *JournalServer) autoFlush(
+	tlfID TlfID, afs JournalAutoFlushStatus,
+	hasWorkCh, pauseCh, resumeCh, shutdownCh chan struct{}) {
+	ctx := ctxWithRandomID(
+		context.Background(), "journal-auto-flush", "1", j.log)
+	for {
+		j.log.CDebugf(ctx, "Waiting for events for %s (%s)", tlfID, afs)
+		switch afs {
+		case JournalAutoFlushEnabled:
+			select {
+			case <-hasWorkCh:
+				j.log.CDebugf(
+					ctx, "Got work event for %s", tlfID)
+				err := j.Flush(ctx, tlfID)
+				if err != nil {
+					j.log.CWarningf(ctx,
+						"Error when flushing %s: %v",
+						tlfID, err)
+				}
+
+			case <-pauseCh:
+				j.log.CDebugf(ctx,
+					"Got pause event for %s", tlfID)
+				afs = JournalAutoFlushDisabled
+
+			case <-shutdownCh:
+				j.log.CDebugf(ctx,
+					"Got shutdown event for %s", tlfID)
+				return
+			}
+
+		case JournalAutoFlushDisabled:
+			select {
+			case <-resumeCh:
+				j.log.CDebugf(ctx,
+					"Got resume event for %s", tlfID)
+				afs = JournalAutoFlushEnabled
+
+			case <-shutdownCh:
+				j.log.CDebugf(ctx,
+					"Got shutdown event for %s", tlfID)
+				return
+			}
+
+		default:
+			j.log.CErrorf(
+				ctx, "Unknown JournalAutoFlushStatus %s", afs)
+			return
+		}
+	}
+}
+
+// PauseAutoFlush pauses the background auto-flush goroutine, if it's
+// not already paused.
+func (j *JournalServer) PauseAutoFlush(ctx context.Context, tlfID TlfID) {
+	j.log.CDebugf(ctx, "Signaling pause for %s", tlfID)
+	bundle, ok := j.getBundle(tlfID)
+	if !ok {
+		j.log.CDebugf(ctx,
+			"Could not find bundle for %s; dropping pause signal",
+			tlfID)
+	}
+
+	select {
+	case bundle.pauseCh <- struct{}{}:
```

Comment:
This is fine, I think, since the channel is buffered for one item. That just means that multiple pause requests (or resume requests) in a row will get coalesced into one, which is fine, since those operations are just meant for ad hoc testing. Specifically, if you do a pause, resume, pause in quick succession, the second pause might be dropped, but I think that's fine. Although come to think of it, if it's in the middle of flushing a long list of requests, it should stop in the middle instead of waiting for the whole list. I'll do that.

I'll also add some comments and change some names to make it clearer what's going on.